stacker/blueprints/base.py

```diff
@@ -152,6 +173,95 @@ def setup_parameters(self):
             p = build_parameter(param, attrs)
             t.add_parameter(p)
 
+    def defined_parameters(self):
+        """Return a dictionary of parameters defined by the blueprint.
+
+        By default, this will just return the values from
+        `BLUEPRINT_PARAMETERS`, but this makes it easy for subclasses to add
+        parameters.
+
+        Returns:
+            dict: parameters defined by the blueprint
+
+        """
+        return getattr(self, "BLUEPRINT_PARAMETERS", {})
+
+    def get_parameters(self):
+        """Return a dictionary of parameters available to the template.
+
+        These parameters will have been defined within `BLUEPRINT_PARAMETERS`
+        or `self.defined_parameters`. Any parameter value that depends on stack
+        output will have been resolved.
+
+        Returns:
+            dict: parameters available to the template
+
+        """
+        if self.resolved_parameters is None:
+            raise UnresolvedBlueprintParameters(self)
+        return self.resolved_parameters
+
+    def resolve_parameters(self, parameter_values, provider, context):
+        """Resolve the values of the blueprint parameters.
+
+        resolve_parameters will resolve the values of the
+        `BLUEPRINT_PARAMETERS` with values from the command lie, the env file,
+        the config, and any references to other stack output.
+
+        Args:
+            parameter_values (dict): dictionary of parameter key to parameter
+                value
+            provider (stacker.Provider): provider to use to resolve references
+                to output
+            context (stacker.Context): context to pull stack fqn from
+
+        """
+
+        def _resolve_string(value, provider, context):
+            for match in OUTPUT_REGEX.finditer(value):
+                # Get from the Output(s) of another stack(s) in the
+                # stack_map
+                stack_output = match.group()
+                stack_name, output = stack_output.split("::")
+                stack_fqn = context.get_fqn(stack_name)
+                try:
+                    output_value = provider.get_output(stack_fqn, output)
+                except KeyError:
+                    raise OutputDoesNotExist(stack_fqn, stack_output)
+                value = OutputTemplate(value).safe_substitute(
```

Comment:
It took me a second to understand why you do the safe_substitute followed by the replace.  Might be worth commenting.